fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -1338,22 +1342,29 @@ public boolean shouldAllowUniquePendingState(FDBRecordStore store) {
          */
         public boolean shouldAllowTypeConversionContinue(IndexBuildProto.IndexBuildIndexingStamp.Method newMethod,
                                                          IndexBuildProto.IndexBuildIndexingStamp.Method oldMethod) {
-            if (allowedTakeoverSet != null) {
-                if (newMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
-                    if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MULTI_TARGET_BY_RECORDS) {
-                        return allowedTakeoverSet.contains(TakeoverTypes.MULTI_TARGET_TO_SINGLE);
-                    }
-                    if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MUTUAL_BY_RECORDS) {
-                        return allowedTakeoverSet.contains(TakeoverTypes.MUTUAL_TO_SINGLE);
-                    }
+            if (newMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
+                if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MULTI_TARGET_BY_RECORDS) {
+                    return isTypeConversionAllowed(TakeoverTypes.MULTI_TARGET_TO_SINGLE);
+                }
+                if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MUTUAL_BY_RECORDS) {
+                    return isTypeConversionAllowed(TakeoverTypes.MUTUAL_TO_SINGLE);
                 }
-                return false;
             }
-            return allowTakeoverContinue &&
-                   // Takeover is allowed only in certain cases
-                   (newMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS &&
-                    (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MULTI_TARGET_BY_RECORDS ||
-                     oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MUTUAL_BY_RECORDS));
+            if (newMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MUTUAL_BY_RECORDS) {
+                if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MULTI_TARGET_BY_RECORDS ||
+                        oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
+                    // Note that single-target-mutual is a private case of multi-target-mutual
+                    return isTypeConversionAllowed(TakeoverTypes.BY_RECORDS_TO_MUTUAL);
+                }
+            }
+            return false;
+        }
+
+        private boolean isTypeConversionAllowed(TakeoverTypes takeoverType) {
+            return
+                    allowedTakeoverSet == null ?
+                    allowTakeoverContinue :
```

Comment:
Perhaps it would make sense to push this into the builder, and have:
```
            public Builder allowTakeoverContinue(boolean allow) {
                this.allowedTakeoverSet = allow ? EnumSet.allOf(TakeoverTypes.class) : EnumSet.noneOf(TakeoverTypes.class);
                return this;
            }
```
and change the default value of `this.allowedTakeoverSet` to `EnumSet.noneOf(TakeoverTypes.class)` instead of `null`

Then this method just becomes `allowedTakeoverSet.contains(takeoverType)`, and may make sense to inline.

Perhaps better as a followup PR though.

Reply:
✅ 

(It was an easy change)